csharp/src/Apache.Arrow/C/CArrowArrayImporter.cs

```diff
@@ -0,0 +1,277 @@
+        /// </remarks>
+        /// <examples>
+        /// Typically, you will allocate an uninitialized CArrowArray pointer,
+        /// pass that to external function, and then use this method to import
+        /// the result.
+        /// 
+        /// <code>
+        /// CArrowArray* importedPtr = CArrowArray.Create();
+        /// foreign_export_function(importedPtr);
+        /// IArrowArray importedArray = CArrowArrayImporter.ImportArray(importedPtr);
+        /// </code>
+        /// </examples>
+        public static unsafe IArrowArray ImportArray(CArrowArray* ptr, IArrowType type)
+        {
+            ImportedArrowArray importedArray = null;
+            try
+            {
+                importedArray = new ImportedArrowArray(ptr);
+                return importedArray.GetAsArray(type);
+            }
+            finally
+            {
+                importedArray?.Release();
+            }
+        }
+
+        /// <summary>
+        /// Import C pointer as a <see cref="RecordBatch"/>.
+        /// </summary>
+        /// <remarks>
+        /// This will call the release callback once all of the buffers in the returned
+        /// RecordBatch are disposed.
+        /// </remarks>
+        /// <examples>
+        /// Typically, you will allocate an uninitialized CArrowArray pointer,
+        /// pass that to external function, and then use this method to import
+        /// the result.
+        /// 
+        /// <code>
+        /// CArrowArray* importedPtr = CArrowArray.Create();
+        /// foreign_export_function(importedPtr);
+        /// RecordBatch batch = CArrowArrayImporter.ImportRecordBatch(importedPtr, schema);
+        /// </code>
+        /// </examples>
+        public static unsafe RecordBatch ImportRecordBatch(CArrowArray* ptr, Schema schema)
+        {
+            ImportedArrowArray importedArray = null;
+            try
+            {
+                importedArray = new ImportedArrowArray(ptr);
+                return importedArray.GetAsRecordBatch(schema);
+            }
+            finally
+            {
+                importedArray?.Release();
+            }
+        }
+
+        private sealed unsafe class ImportedArrowArray : ImportedAllocationOwner
+        {
+            private readonly CArrowArray* _cArray;
+
+            public ImportedArrowArray(CArrowArray* cArray)
+            {
+                if (cArray == null)
+                {
+                    throw new ArgumentNullException(nameof(cArray));
+                }
+                _cArray = cArray;
+                if (_cArray->release == null)
+                {
+                    throw new ArgumentException("Tried to import an array that has already been released.", nameof(cArray));
+                }
+            }
+
+            protected override void FinalRelease()
+            {
+                if (_cArray->release != null)
+                {
+                    _cArray->release(_cArray);
```

Comment:
Do you need to free `_cArray`?

Nevermind.  I see that `_cArray` comes from the caller of `ImportArray` and they are responsible for deleting it (which it seems they can always do as soon as the call to `ImportArray` is finished).